compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -1608,134 +1609,122 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         let param_env = tcx.param_env(block.owner);
         let cause = ObligationCause::misc(span, block.owner.def_id);
 
-        let mut fulfillment_errors = Vec::new();
-        let mut applicable_candidates: Vec<_> = infcx.probe(|_| {
-            // Regions are not considered during selection.
-            let self_ty = self_ty
-                .fold_with(&mut BoundVarEraser { tcx, universe: infcx.create_next_universe() });
-
-            struct BoundVarEraser<'tcx> {
-                tcx: TyCtxt<'tcx>,
-                universe: ty::UniverseIndex,
-            }
-
-            // FIXME(non_lifetime_binders): Don't assign the same universe to each placeholder.
-            impl<'tcx> TypeFolder<TyCtxt<'tcx>> for BoundVarEraser<'tcx> {
-                fn interner(&self) -> TyCtxt<'tcx> {
-                    self.tcx
-                }
+        let mut universes = if self_ty.has_escaping_bound_vars() {
+            vec![None; self_ty.outer_exclusive_binder().as_usize()]
+        } else {
+            vec![]
+        };
 
-                fn fold_region(&mut self, r: ty::Region<'tcx>) -> ty::Region<'tcx> {
-                    // FIXME(@lcnr): This is broken, erasing bound regions
-                    // impacts selection as it results in different types.
-                    if r.is_bound() { self.tcx.lifetimes.re_erased } else { r }
-                }
+        crate::traits::project::with_replaced_escaping_bound_vars(
+            infcx,
+            &mut universes,
+            self_ty,
+            |self_ty| {
+                let tcx = self.tcx();
+                let InferOk { value: self_ty, obligations } =
+                    infcx.at(&cause, param_env).normalize(self_ty);
+
+                let (impl_, (assoc_item, def_scope)) = self.select_inherent_assoc_type_candidates(
+                    infcx,
+                    name,
+                    span,
+                    self_ty,
+                    cause,
+                    param_env,
+                    obligations,
+                    candidates,
+                )?;
```

Comment:
I may be throwing a wrench into this problem since I haven't been keeping up with the investigation, but why does this even have side-effects in the first place?

IMO, the equate and select should be done in a **probe**, because any side-effects of inherent type projection be [re-confirmed](https://github.com/rust-lang/rust/blob/beebcdeb6fc778c2d082d545554b08183121b42e/compiler/rustc_trait_selection/src/solve/project_goals/weak_types.rs#L20-L30) later when the type is actually projected.

If we do this in a probe, that means that the substs that we return back below (parent args) can be prepared *outside* of the `with_replaced_escaping_bound_vars`, and *don't* need to be resolved eagerly or anything like that.

The `with_replaced_escaping_bound_vars` can just return a `bool` (or `Result<(), NoSolution>`) to signify if the probe was successful.

For example:

```
impl Foo<i32> {
  type Assoc = i32;
}
```
If we try to select `<Foo<_>>::Assoc`, then we should *return* `<Foo<?0>>::Assoc`, not `<Foo<i32>>::Assoc`, because we can leave the `?0` to be constrained during projection.

This is more similar to how GAT selection is done, which also has no side-effects.

Reply:
The following diff works btw ...

```diff
diff --git a/compiler/rustc_hir_analysis/src/astconv/mod.rs b/compiler/rustc_hir_analysis/src/astconv/mod.rs
index 1c841746211..43e705f5d1a 100644
--- a/compiler/rustc_hir_analysis/src/astconv/mod.rs
+++ b/compiler/rustc_hir_analysis/src/astconv/mod.rs
@@ -26,14 +26,13 @@
 use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_hir::intravisit::{walk_generics, Visitor as _};
 use rustc_hir::{GenericArg, GenericArgs, OpaqueTyOrigin};
-use rustc_infer::infer::resolve::EagerResolver;
 use rustc_infer::infer::{InferCtxt, InferOk, TyCtxtInferExt};
 use rustc_infer::traits::{Obligation, ObligationCause};
 use rustc_middle::middle::stability::AllowUnstable;
 use rustc_middle::ty::GenericParamDefKind;
 use rustc_middle::ty::{
     self, Const, GenericArgKind, GenericArgsRef, IsSuggestable, ParamEnv, Predicate, Ty, TyCtxt,
-    TypeFoldable, TypeVisitableExt,
+    TypeVisitableExt,
 };
 use rustc_session::lint::builtin::AMBIGUOUS_ASSOCIATED_ITEMS;
 use rustc_span::edit_distance::find_best_match_for_name;
@@ -1621,19 +1620,22 @@ fn lookup_inherent_assoc_ty(
             self_ty,
             |self_ty| {
                 let tcx = self.tcx();
-                let InferOk { value: self_ty, obligations } =
-                    infcx.at(&cause, param_env).normalize(self_ty);
 
-                let (impl_, (assoc_item, def_scope)) = self.select_inherent_assoc_type_candidates(
-                    infcx,
-                    name,
-                    span,
-                    self_ty,
-                    cause,
-                    param_env,
-                    obligations,
-                    candidates,
-                )?;
+                let (impl_, (assoc_item, def_scope)) = infcx.probe(|_| {
+                    let InferOk { value: self_ty, obligations } =
+                        infcx.at(&cause, param_env).normalize(self_ty);
+
+                    self.select_inherent_assoc_type_candidates(
+                        infcx,
+                        name,
+                        span,
+                        self_ty,
+                        cause,
+                        param_env,
+                        obligations,
+                        candidates,
+                    )
+                })?;
 
                 self.check_assoc_ty(assoc_item, name, def_scope, block, span);
 
@@ -1649,7 +1651,6 @@ fn lookup_inherent_assoc_ty(
                 );
 
                 let ty = Ty::new_alias(tcx, ty::Inherent, ty::AliasTy::new(tcx, assoc_item, args));
-                let ty = ty.fold_with(&mut EagerResolver::new(infcx));
 
                 Ok(Some((ty, assoc_item)))
             },
diff --git a/compiler/rustc_trait_selection/src/traits/project.rs b/compiler/rustc_trait_selection/src/traits/project.rs
index 8f5afdf0a1f..aed35b3343f 100644
--- a/compiler/rustc_trait_selection/src/traits/project.rs
+++ b/compiler/rustc_trait_selection/src/traits/project.rs
@@ -1455,7 +1455,11 @@ pub fn compute_inherent_assoc_ty_args<'a, 'b, 'tcx>(
     // Infer the generic parameters of the impl by unifying the
     // impl type with the self type of the projection.
     let self_ty = alias_ty.self_ty();
-    match selcx.infcx.at(&cause, param_env).eq(DefineOpaqueTypes::No, impl_ty, self_ty) {
+    let infcx = selcx.infcx.at(&cause, param_env);
+    let InferOk { value: self_ty, obligations: normalize_obligations } = infcx.normalize(self_ty);
+    obligations.extend(normalize_obligations);
+
+    match infcx.eq(DefineOpaqueTypes::No, impl_ty, self_ty) {
         Ok(mut ok) => obligations.append(&mut ok.obligations),
         Err(_) => {
             tcx.sess.delay_span_bug(
```

As @compiler-errors said "I think there's a trade-off that needs to be decided here. I'm leaning towards IAT selection having no side effects, but that does mean that we need to move normalization to WF, because right now the code is normalizing during astconv, which seems bad imo"

Reply:
For the record, that diff is not the most efficient way to do this.

The best way to do apply the changes I asked for is just to move the normalize call into `select_inherent_assoc_type_candidates ` and use `ocx.normalize`. And also, in wf (`compute_inherent_assoc_ty_args `), you'd just want to use `normalize_with_depth` like is happening immediately above the code you touched.

Here's a diff that shows these changes: https://github.com/rust-lang/rust/compare/master...compiler-errors:rust:do-not-erase?expand=1